### PR TITLE
#10 Add Subscription support

### PR DESCRIPTION
**Pull request by Allali84**
Hello,

Here is my proposal for the adding of subscription support

Best regards,

### CODE REVIEW

File: src/main/kotlin/com/apurebase/kgraphql/schema/Subscription.kt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.apurebase.kgraphql.schema
```
Comment (Allali84) [inline src/main/kotlin/com/apurebase/kgraphql/schema/Subscription.kt]: Sorry I forgot to remove this class :P

File: src/main/kotlin/com/apurebase/kgraphql/schema/model/MutableSchemaDefinition.kt [hunk 2]
```diff
@@ -109,6 +110,13 @@ data class MutableSchemaDefinition (
                 mutations.add(mutation)
         }
 
+        fun addSubscription(subscription: SubscriptionDef<*>){
+                if(subscription.checkEqualName(subscriptions)){
+                        throw SchemaException("Cannot add mutation with duplicated name ${subscription.name}")
```
Comment (Allali84) [inline src/main/kotlin/com/apurebase/kgraphql/schema/model/MutableSchemaDefinition.kt]: I need to change the message in this exception :P

**Comment by Allali84** [comment]
I already start a project with spring boot web-services ([spring boot starter for kgraphql](https://github.com/Allali84/kgraphql-spring-boot)) and i m waiting for the subscription feature to add it to my project too, for implementing it i'm thinking about the reactive web flux of spring
I never worked before with ktor, but i'm really considering it, i think i'm really interested on working on the #37

**Comment by jeggy** [comment]
Could you fix the small conflicts, and then we'll merge this as of now :)

**Comment by Allali84** [comment]
it's done :)

**Comment by nidomiro** [comment]
Hey @Allali84 , thanks for implementing Subscriptions :)

Currently I'm reading into the library to be able to help with developing.
While reading through your changes, I hard time understanding how to use the implementation of the subscription. 
I hope you can help me understand ;)

The thing I would expect when executing a subscription is a continuous stream of elements.
Example:
```
schema.execute("subscription { numbers }").forEach(::println)
// Output
1
2
...
```
I know it is not easy to implement this way, but that's the thing I would expect.
The execute method would give back something like a "coroutine iterator".

**Comment by Allali84** [comment]
Hello @nidomiro, you are welcome :)
I updated my project ([spring boot starter for kgraphql](https://github.com/Allali84/kgraphql-spring-boot)) to add an example of how to use this subscription, i hope it can help you understand it
Feel free to ask me if it always ambiguous for you :)

**Comment by nidomiro** [comment]
Thank you for the Example.
If I understand it correctly your initiate the subscription via one endpoint and it will be available on a completely new endpoint?
If that's the case you would have a GraphQL-Rest-hybrid.

I already used https://github.com/graphql-java-kickstart/graphql-spring-boot and there it would be like the following:
1. you establish a websocket to `/subscriptions`
2. send your subscription query
3. receive updated values as response as long as the connection is open

**Comment by Allali84** [comment]
I think it's a way of doing things.
I believe the advantage of what I did that you can use the same subscription from multiple client.

1- subscribe and give it a name for example `MySubscription` using `/graphql`
2- use websocket `/subscription/MySubscription` from any client to receive updates

**Comment by nidomiro** [comment]
You can do the same with my example

**Comment by jeggy** [comment]
I was maybe too fast on merging this, but I think this is fine for now.

Which graphql clients do you guys use? My first goal is to add support for the [apollo-client - Subscriptions](https://www.apollographql.com/docs/react/data/subscriptions/).


I started looking at other implementations of Subscription. Apollo uses this one: https://github.com/apollographql/graphql-subscriptions

We could do a implementation like that one, this way we don't limit subscribing to a single mutation, but we can subscribe to any events. 

This is just some idea spitting:

```kotlin
data class MyNormalValueClass(val value: String)
class MyTypeEvent(val value: MyNormalValueClass) : PubSubEventSealedClass()

...

val pubsub = PubSub<MyTypeEvent, MyNormalValue> { event: MyTypeEvent ->
  // This callback would be called once every event and sent to all subscriptions
  event.value
}


mutation("add") {
  resolver<MyNormalValueClass> { value: String ->
    val returnValue = MyNormalValueClass(value)
    pubsub.publish(MyTypeEvent(returnValue))
    returnValue
  }
}

subscription<MyNormalValueClass>("listenToAdds") {
  pubsub.iterator()
}
```

**Comment by nidomiro** [comment]
Yes, it's a base to build on.
But I think we should move and continue the conversation in a ticket (#10).
The main points (in my opinion) are moved to the ticket.
@jeggy @Allali84: feel free to add your opinion ;)

**Pull request merged**